main.cpp

```diff
@@ -19,13 +19,34 @@ using namespace clang::tooling;
 
 class CastCallBack : public MatchFinder::MatchCallback {
 public:
-    CastCallBack(Rewriter& rewriter) {
-        // Your code goes here
+    CastCallBack(Rewriter& rewriter) : rewriter_(rewriter) {
     };
 
     void run(const MatchFinder::MatchResult &Result) override {
-        // Your code goes here
+    const auto* result_item = Result.Nodes.getNodeAs<CStyleCastExpr>("cast");
+	SourceManager& source_manager = *Result.SourceManager;
+
+	auto re_range = CharSourceRange::getCharRange (result_item->getLParenLoc(), result_item->getSubExprAsWritten()->getBeginLoc());
+
+	StringRef type_string = 
+              Lexer::getSourceText(CharSourceRange::getTokenRange(result_item->getLParenLoc().getLocWithOffset(1),
+                                   result_item->getRParenLoc().getLocWithOffset(-1)),
+		                           source_manager, Result.Context->getLangOpts());
+
+	std::string str = ("static_cast<" + type_string + ">").str();
+
+	const Expr* sub_expr = result_item->getSubExprAsWritten()->IgnoreImpCasts();
```

Comment:
Please describe what happens if change the code as follows:  
```suggestion
	const Expr* sub_expr = result_item->getSubExpr();
```
What is the difference between `getSubExprAsWritten()` and `getSubExpr()`? Please provide an example.  
What is the purpose of using the `IgnoreImpCasts()` method?

Reply:
1)`const Expr* sub_expr = result_item->getSubExprAsWritten()->IgnoreImpCasts();`
Output: `int main() {
    float f;
    int i = static_cast<int>(f);
    return 0;
}`
`const Expr* sub_expr = result_item->getSubExpr();`
Output: `int main() {
    float f;
    int i = static_cast<int>(f);
    return 0;
}`
Ничего не изменилось c:
3) `IgnoreImpCasts()` пропускает неявное приведение типов, здесь использовать не обязательно.
2) `getSubExprAsWritten()` Retrieve the cast subexpression as it was written in the source code, looking through any implicit casts or other intermediate nodes introduced by semantic analysis.
`getSubExpr()` take expr after semantic analysis  `->` result with `getSubExpr()` may be incorrect in the case of semantic analysis)

Reply:
If you try to print the AST of the `test.cpp` example, you will see that you have several `ImplicitCastExpr` nodes:
![image](https://user-images.githubusercontent.com/76072618/119860568-20cb0700-bf1f-11eb-8e3d-2a533baa265f.png)

Depending on which approaches you use to get the subexpression node, you will get different result nodes for the `test.cpp` example:

Code:
```cpp
// getSubExpr() retrieves the first cast subexpression
llvm::outs() << "getSubExpr(): " << castNode->getSubExpr()->getStmtClassName();

// getSubExprAsWritten() retrieves the cast subexpression as it was written in the source
// code, looking through any implicit casts or other intermediate nodes
// introduced by semantic analysis.
llvm::outs() << "getSubExprAsWritten(): " << castNode->getSubExprAsWritten()->getStmtClassName();

// getSubExpr() retrieves the first cast subexpression.
// ignoreImpCasts() skips past any implicit casts which might surround this expression until
// reaching a fixed point. It skips the following AST nodes: ImplicitCastExpr, FullExpr.
llvm::outs() << "getSubExpr()->ignoreImpCasts(): " << castNode->getSubExpr()->IgnoreImpCasts()->getStmtClassName();
```
Output:
```bash
 getSubExpr(): ImplicitCastExpr 
 getSubExprAsWritten(): DeclRefExpr  
 getSubExpr()->ignoreImpCasts(): DeclRefExpr
```